### PR TITLE
Add logout SMS method

### PR DESCRIPTION
**Pull request by Jeasmine**
* Add logoutSMS() public method
* Add logout SMS functionality to UserStateSynchronizers
* Remove parent player update from SMS
* Update tests after removal of parent player id link from SMS
* Add logoutSMS tests

## 
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275)

### CODE REVIEW

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java [hunk 1]
```diff
@@ -321,6 +331,49 @@ void onSuccess(String response) {
...
+
+            jsonBody.put(APP_ID, syncValues.optString(APP_ID));
+        } catch (JSONException e) {
+            e.printStackTrace();
+        }
+
+        OneSignalRestClient.postSync(urlStr, jsonBody, new OneSignalRestClient.ResponseHandler() {
+            @Override
+            void onFailure(int statusCode, String response, Throwable throwable) {
+                OneSignal.Log(OneSignal.LOG_LEVEL.WARN, "Failed last request. statusCode: " + statusCode + "\nresponse: " + response);
```
Comment (aimiebell) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java:357]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/SynchronizerIntegrationTests.java [hunk 2]
```diff
@@ -631,7 +631,7 @@ public void shouldFireOnFailureOfEmailAndSMSlUpdateOnNetworkFailure() throws Exc
     }
 
     @Test
-    public void shouldFireOnSuccessOnlyAfterNetworkCallAfterLogout() throws Exception {
+    public void shouldFireOnSuccessOnlyAfterNetworkCallAfterEmailLogout() throws Exception {
         OneSignalInit();
```
Comment (aimiebell) [inline OneSignalSDK/unittest/src/test/java/com/test/onesignal/SynchronizerIntegrationTests.java:635]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

**Review by jkasten2 — CHANGES_REQUESTED** [review]
> Reviewed 4 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275)* status: 4 of 6 files reviewed, 4 unresolved discussions (waiting on @emawby and @Jeasmine)

___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalStateSynchronizer.java, line 248 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUhYpN2AOW7pV7TNqc4:-MUhYpN2AOW7pV7TNqc5:b-rk4bzl) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalStateSynchronizer.java#L248)):*
> ```Java
> 
>    static void logoutSMS() {
>       getPushStateSynchronizer().logoutSMS();
> ```

This seems line seems wrong to me, why are we calling `logoutSMS()` on push?
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java, line 263 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUhZXYZ4g_sOgcAdxIG:-MUhZXYZ4g_sOgcAdxIH:bql55le) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java#L263)):*
> ```Java
>         }
> 
>         if (syncSMSLogout() && userId != null) {
> ```

These `syncEmailLogout` and `syncSMSLogout()` checks seems very out of place to me in `UserStateSynchronizer.java`. This was existing logic before this project so no need to clean up in this PR.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java, line 335 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUh_NDa2AiXI8MOUqm5:-MUh_NDa2AiXI8MOUqm6:b-ugir7w) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java#L335)):*
> ```Java
> 
>     private void doSMSLogout(String userId) {
>         String urlStr = "players/" + userId + "/sms_logout";
> ```

We won't need to make a `sms_logout` call to the backend. By logout we are just no longer updating the SMS record from the SDK. I'll write up some details on this in a spec as a follow up too
___
*[OneSignalSDK/unittest/src/test/java/com/test/onesignal/SynchronizerIntegrationTests.java, line 1844 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUhXJtU44dR7dH3VoPx:-MUhXJtU44dR7dH3VoPy:b-m6b4kt) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/unittest/src/test/java/com/test/onesignal/SynchronizerIntegrationTests.java#L1844)):*
> ```Java
>         threadAndTaskWait();
> 
>         // 4. Attempt a set external user id with callback
> ```

nit, should be 3.

**Review by Jeasmine — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275)* status: 4 of 6 files reviewed, 4 unresolved discussions (waiting on @emawby and @jkasten2)

___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalStateSynchronizer.java, line 248 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUhYpN2AOW7pV7TNqc4:-MUsoLDc9XdlrPUFyLrI:b-rk9vwt) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalStateSynchronizer.java#L248)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

This seems line seems wrong to me, why are we calling `logoutSMS()` on push?
</blockquote></details>

This was emulating email behavior, but now I understand that email was wrong too. The logout logic could be done under each Synchronizers channel
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java, line 263 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUhZXYZ4g_sOgcAdxIG:-MUsmAja941kWOiGQaRS:b-uu9rdw) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java#L263)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

These `syncEmailLogout` and `syncSMSLogout()` checks seems very out of place to me in `UserStateSynchronizer.java`. This was existing logic before this project so no need to clean up in this PR.
</blockquote></details>

Agree. This is one of the differences with iOS. We wait for the push channel to logout email and SMS to update email and SMS channels. I can improve this code in another PR.

Ticket for following this up https://app.asana.com/0/1193807006058450/1200001538876686
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java, line 335 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUh_NDa2AiXI8MOUqm5:-MUsm8Gn9Sr2bkIBvkn1:b-vzs1sk) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java#L335)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

We won't need to make a `sms_logout` call to the backend. By logout we are just no longer updating the SMS record from the SDK. I'll write up some details on this in a spec as a follow up too
</blockquote></details>

gotcha!
___
*[OneSignalSDK/unittest/src/test/java/com/test/onesignal/SynchronizerIntegrationTests.java, line 1844 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUhXJtU44dR7dH3VoPx:-MUsp2Jb8cd8BU8mN_9X:b-5qioa2) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/unittest/src/test/java/com/test/onesignal/SynchronizerIntegrationTests.java#L1844)):*
> ```
>  4. Attempt a set external user id with callback
> ```
Nice catch!

**Comment by Jeasmine** [comment]
> > _[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275)_ status: 4 of 6 files reviewed, 4 unresolved discussions (waiting on @emawby and @jkasten2)
> 
> _[OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalStateSynchronizer.java, line 248 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUhYpN2AOW7pV7TNqc4:-MUsoLDc9XdlrPUFyLrI:b-rk9vwt) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalStateSynchronizer.java#L248)):_
> 
> Previously, jkasten2 (Josh Kasten) wrote…
> This was emulating email behavior, but now I understand that email was wrong too. The logout logic could be done under each Synchronizers channel
> 
> _[OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java, line 263 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUhZXYZ4g_sOgcAdxIG:-MUsmAja941kWOiGQaRS:b-uu9rdw) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java#L263)):_
> 
> Previously, jkasten2 (Josh Kasten) wrote…
> Agree. This is one of the differences with iOS. We wait for the push channel to logout email and SMS to update email and SMS channels. I can improve this code in another PR.
> 
> Ticket for following this up https://app.asana.com/0/1193807006058450/1200001538876686
> 
> _[OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java, line 335 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUh_NDa2AiXI8MOUqm5:-MUsm8Gn9Sr2bkIBvkn1:b-vzs1sk) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java#L335)):_
> 
> Previously, jkasten2 (Josh Kasten) wrote…
> gotcha!
> 
> _[OneSignalSDK/unittest/src/test/java/com/test/onesignal/SynchronizerIntegrationTests.java, line 1844 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUhXJtU44dR7dH3VoPx:-MUsp2Jb8cd8BU8mN_9X:b-5qioa2) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/unittest/src/test/java/com/test/onesignal/SynchronizerIntegrationTests.java#L1844)):_
> 
> > ```
> >  4. Attempt a set external user id with callback
> > ```
> 
> Nice catch!

logout SMS call push sync because it need to remove the data from the UserState

**Review by jkasten2 — APPROVED** [review]
> Reviewed 8 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @emawby)

___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java, line 263 at r1](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1275#-MUhZXYZ4g_sOgcAdxIG:-MV-BEYz7Xuf5ifbGHLp:b4c9lyr) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/a8beb4c59ae59946dacd4833a98fe7cc60f55350/OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java#L263)):*
<details><summary><i>Previously, Jeasmine (JNahui) wrote…</i></summary><blockquote>

Agree. This is one of the differences with iOS. We wait for the push channel to logout email and SMS to update email and SMS channels. I can improve this code in another PR.

Ticket for following this up https://app.asana.com/0/1193807006058450/1200001538876686
</blockquote></details>

:thumbsup:

**Pull request merged**